design/service-discovery.md

```diff
@@ -0,0 +1,178 @@
+
+Using `properties` instead of `USER_DATA` would be preferable in this case, but unfortunately `properties` are not available at publisher/subscriber level on Fast DDS at this moment.
+
+## Class diagram ##
+
+The following class diagram only shows classes modified by this solution. And for those classes, only added members (`+`), modified (`~`) or removed (`-`) members are shown.
+
+![Class diagram](service-discovery-images/class-diagram.svg)
+
+## Server side ##
+
+On the server side the general idea is to hold the incoming requests until the response publisher has matched with the response subscriber **that corresponds to the request publisher** sending the request.
+That is, the request will not be notified to the upper layer until the response channel is fully established.
+
+When a new request publisher is discovered, the GUID of the corresponding response subscriber is taken from the `USER_DATA` and the mapping between the request publisher and response subscriber is stored.
+Also, when creating the request subscriber, the response publisher's GUID is added to the `USER_DATA` for the client to be able to make the connection between them.
+
+### rmw_create_service ###
+
+This method creates the request subscriber first and the response publisher afterwards.
+The order of the creations must be inverted, so that we can get the publisher's GUID and store it on the `USER_DATA` on the `subscriberParam` before creating the subscriber.
+
+### CustomParticipantInfo ###
+
+* Add `remote_response_guid_by_remote_request_guid_` to hold the relation between a request publisher and its corresponding response subscriber.
+  It will be an unordered map with the request publisher GUID as key:
+  `std::unordered_map<eprosima::fastrtps::rtps::GUID_t, eprosima::fastrtps::rtps::GUID_t, rmw_fastrtps_shared_cpp::hash_fastrtps_guid>`
+
+* Add method `remote_response_guid` to retrieve the content of `remote_response_guid_by_remote_request_guid_` given the GUID of a request endpoint. If the given GUID is not among the keys of the map, it returns `c_Guid_Unknown`.
+
+![CustomParticipantInfo](service-discovery-images/CustomParticipantInfo.svg)
+
+### ParticipantListener::constructor ###
+
+The constructor needs to be modified to accept a `CustomParticipantInfo` parameter, for the listener callbacks to have access to the `remote_response_guid_by_remote_request_guid_`.
+
+### ParticipantListener::onPublisherDiscovery ###
+
+* When a new publisher is discovered, read the `USER_DATA` to get the corresponding response subcriber's GUID and store the relation on the `remote_response_guid_by_remote_request_guid_` map.
+* If the `USER_DATA` dos not contain a GUID, **do not add an entry to the map**. This signals that the remote endpoint is not compiant with these modifications (for backwards compatibility and with other vendors).
+* When a publisher is un-discovered, remove the entry from the `remote_response_guid_by_remote_request_guid_` map (if any).
```

Comment:
That somehow the subscriber stops considering it as a matched peer. It can be that the peer disconnects or that it does not assert its liveliness. In any case, the subscriber will consider that the publisher is not reachable.

Reply:
Would you consider replacing `un-discovered` by `no longer reachable`?